Sprint 10

tayawarmly opened the pull request:
- Отображение прогресса загрузки веб-страницы авторизации (KVO)
- Добавлен Splash Screen
- Добавлена структура для декодинга JSON-ответа от POST-запроса авторизации Unsplash
- Добавлено логирование ошибок в консоль

tayawarmly committed:
- Авторизация через OAuth2.0
- Отображение прогресса загрузки веб-страницы авторизации (KVO)
- Добавлен Splash Screen
- Добавлена структура для декодинга JSON-ответа от POST-запроса авторизации Unsplash
- Добавлено логирование ошибок в консоль

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,55 @@
+    func webViewViewControllerDidCancel(_ vc: WebViewViewController) {
+        dismiss(animated: true)
+    }
+}
```
В этом классе все отлично, молодец!

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,66 @@
+final class OAuth2Service {
+    static let shared = OAuth2Service()
+    private let urlSession = URLSession.shared
+    private let decoder = JSONDecoder()
```
Круто, что декодер вынеста как свойство, чтобы каждый раз новый экземпляр не создавать, когда вызывается свойство `fetchOAuthToken`

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,66 @@
+    
+    private init() {}
+    
+    func makeOAuthTokenRequest(code: String) -> URLRequest {
```
Так как метод используется только внутри класса, давай его инкапсулируем и используем ключевое слово `private`

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,66 @@
+    
+    func makeOAuthTokenRequest(code: String) -> URLRequest {
+        guard let baseURL = URL(string: "https://unsplash.com") else {
+            preconditionFailure("Invalid baseURL")
```
Зачем выбрасывать такую ошибку ? Почему бы не сделать метод с опциональным возвращаемым значением:
```
private func makeOAuthTokenRequest(code: String) -> URLRequest?
```
И в случае ошибки просто возвращать `nil` ?
```
        guard let baseURL = URL(string: "https://unsplash.com") else { return nil }
        let url = URL(
            string: "/oauth/token"
            + "?client_id=\(Constants.accessKey)"
            + "&&client_secret=\(Constants.secretKey)"
            + "&&redirect_uri=\(Constants.redirectURI)"
            + "&&code=\(code)"
            + "&&grant_type=authorization_code",
            relativeTo: baseURL
        )

        guard let url else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        return request
```
В методе `fetch` все еще проще:
```
    func fetchOAuthToken (code: String, completion: @escaping (Result<String, Error>) -> Void) {
        guard let request = makeOAuthTokenRequest(code: code) else { return }
```
Но я бы рекомендовал обработать ошибку там:
```
    enum ServiceErrors: Error {
        case makeTokenRequestError
    }

    func fetchOAuthToken (code: String, completion: @escaping (Result<String, Error>) -> Void) {
        guard let request = makeOAuthTokenRequest(code: code) else {
            completion(.failure(ServiceErrors.makeTokenRequestError))
            return
        }
```

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,66 @@
+        
+        let task = urlSession.data(for: request) { [weak self] result in
+            guard let self else {
+                preconditionFailure("fetchOAuthToken error: self is nil")
```
В данном случае нужно обработать ошибку:
```
            guard let self else {
                completion(.failure(ServiceErrors.selfError))
                return
            }
```

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,66 @@
+                do {
+                    decoder.keyDecodingStrategy = .convertFromSnakeCase
+                    let response = try decoder.decode(OAuthTokenResponseBody.self, from: data)
+                    print(response)
```
Print для дебага можно удалить.

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,66 @@
+                    
+                    completion(.success(response.accessToken))
+                } catch {
+                    print("Failed to parse data: \(error.localizedDescription)")
```
Аналогично и здесь, давай обработаем ошибку 
```
                } catch {
                    completion(.failure(ServiceErrors.decodingError))
                    print("Failed to parse data: \(error.localizedDescription)")
                }
```

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,66 @@
+                }
+                
+            case .failure(let error):
+                completion(.failure(error))
```
Обрати внимание, по ТЗ необходимо чтобы во всех `guard`-ах и во всех кейсах `failure` в данном сервисе было распечатано событие ошибки, поэтому пройдись по всем рекомендациям выше и проверь наличие `print("Описание какая ошибка отработала")`

So1ovyov commented on ImageFeed/Auth/URLSession+data.swift:
```diff
@@ -0,0 +1,43 @@
+                if 200 ..< 300 ~= statusCode {
+                    fulfillCompletionOnTheMainThread(.success(data))
+                } else {
+                    fulfillCompletionOnTheMainThread(.failure(NetworkError.httpStatusCode(statusCode)))
```
Ни одна из ошибок не распечатана в терминале, а это необходимо по ТЗ.
Либо тут для каждого кейса `failure` надо добавить `print`, либо сделать это в сервисе в кейсе:
```
            case .failure(let error):
                print("Тут описание ошибки \(error)")
                completion(.failure(error))
```

So1ovyov commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,133 @@
+    
+    @objc
+    private func didTapBackButton() {
+        print("нажатие кнопки назад")
```
Хорошей практикой считается написание `TODO`-шки что будет сделано в следующей итерации

So1ovyov commented on ImageFeed/SplashViewController.swift:
```diff
@@ -0,0 +1,94 @@
+
+import UIKit
+
+class SplashViewController: UIViewController {
```
Проверь во всех классах наличие модификатора `final`, он не нужен только если ты планируешь наследоваться от класса
```
final class SplashViewController: UIViewController {
```

So1ovyov commented on ImageFeed/SplashViewController.swift:
```diff
@@ -0,0 +1,94 @@
+                assertionFailure("Failed to prepare for \(showAuthenticationScreenSegueIdentifier)")
+                return
+            }
+
```
Лишние пустые строки внутри метода можно удалить.

So1ovyov commented on ImageFeed/SplashViewController.swift:
```diff
@@ -0,0 +1,94 @@
+    // переход к галерее
+    func didAuthenticate(_ vc: AuthViewController, didAuthenticateWith code: String) {
+        dismiss(animated: true) { [weak self] in
+            guard let self = self else { return }
```
С определенной версии `swift`, если не ошибаюсь с 5.3 можно делать вот такую конструкцию:
```
guard let self else { return }
```

tayawarmly committed:
- Исправлен переход в галерею, если пользователь уже авторизован в приложении
- Исправлен шрифт текста на кнопке Войти
- Добавлено недостающие логирование
- Исправлены мелкие недочеты

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,84 @@
+
+final class OAuth2Service {
+    static let shared = OAuth2Service()
+    var authToken: String? {
```
В этом нет необходимости, достаточно создать константу:
```
private let storage = OAuth2TokenStorage()
```
И применить её в коде ниже на 68-ой строке:
```
storage.token = response.accessToken
```

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,84 @@
+            + "&&grant_type=authorization_code",
+            relativeTo: baseURL)
+        
+        guard let url else { return nil }
```
Здесь также нужно было добавить `print`
Добавлено логирование ошибок в консоль, распечатаны "Ошибочные кейсы, при получении опциональных значений, таких как URLComponents, URL, URLRequest"

Конкретно здесь не залогирована ошибка при получении URL

So1ovyov commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -0,0 +1,84 @@
+    
+    func fetchOAuthToken (_ code: String, completion: @escaping (Result<String, Error>) -> Void) {
+        guard let request = makeOAuthTokenRequest(code: code) else {
+            completion(.failure(ServiceErrors.makeTokenRequestError))
```
Здесь пропустила `print` в случае ошибки

А здесь не залогирована ошибка при получении URLRequest

tayawarmly replied:
Спасибо за ревью! Добавила недостающие строчки в код)

So1ovyov commented:
@tayawarmly привет! Почти всё исправлено, но ты пропустила 2 места, где необходимо добавить принты в случае ошибок. Не хочется отправлять на еще одну пересдачу, поэтому пожалуйста к следующему спринту добавь принты, там где я указал в коде.
Успехов в следующем спринте!

tayawarmly committed:
.
